gpu-aware-scheduling/pkg/gpuscheduler/scheduler.go

```diff
@@ -0,0 +1,604 @@
+	}
+
+	return annotation, nil
+}
+
+// checkResourceCapacity returns true if the needed resources fit based on capacity and used resources.
+func checkResourceCapacity(neededResources, capacity, used resourceMap) bool {
+	for resName, resNeed := range neededResources {
+		if resNeed < 0 {
+			klog.Error("negative resource request")
+
+			return false
+		}
+
+		resCapacity, ok := capacity[resName]
+		if !ok || resCapacity <= 0 {
+			klog.V(L4).Info(" no capacity available for ", resName)
+
+			return false
+		}
+
+		resUsed := used[resName] // missing = 0, default value is ok
+
+		if resUsed < 0 {
+			klog.Error("negative amount of resources in use")
+
+			return false
+		}
+
+		klog.V(L4).Info(" resource ", resName, " capacity:", strconv.FormatInt(resCapacity, 10), " used:",
+			strconv.FormatInt(resUsed, 10), " need:", strconv.FormatInt(resNeed, 10))
+
+		if resUsed+resNeed < 0 {
+			klog.Error("resource request overflow error")
+
+			return false
+		}
+
+		if resCapacity < resUsed+resNeed {
+			klog.V(L4).Info(" not enough resources")
+
+			return false
+		}
+	}
+
+	klog.V(L4).Info(" there is enough resources")
+
+	return true
+}
+
+func (m *GASExtender) bindNode(args *extender.BindingArgs) *extender.BindingResult {
+	result := extender.BindingResult{}
+
+	// first get the pod
+	pod, err := m.cache.fetchPod(args.PodNamespace, args.PodName)
+	if err != nil {
+		klog.Warningf("Pod %s couldn't be read or pod vanished", args.PodName)
+
+		result.Error = err.Error()
+
+		return &result
+	}
+
+	m.rwmutex.Lock()
+	klog.V(L5).Infof("bind %v:%v to node %v locked", args.PodNamespace, args.PodName, args.Node)
+	defer m.rwmutex.Unlock()
+
+	// pod should always fit, but one never knows if something bad happens between filtering and binding
+	annotation, err := m.getAnnotationForPodGPURequest(pod, args.Node)
+	resourcesAdjusted := false
+
+	if err == nil {
+		klog.V(L3).Infof("bind %v:%v to node %v annotation %v", args.PodNamespace, args.PodName, args.Node, annotation)
+		// adjust (add) resources in cache, this includes sanity check for the inputs
+		err = m.cache.adjustPodResourcesL(pod, add, annotation, args.Node)
+
+		if err == nil {
+			resourcesAdjusted = true
+			// annotate POD with per-container GPU selection
+			err = m.annotatePodBind(annotation, pod)
+
+			if err == nil {
+				// now bind
+				binding := &v1.Binding{
+					ObjectMeta: metav1.ObjectMeta{Name: args.PodName, UID: args.PodUID},
+					Target:     v1.ObjectReference{Kind: "Node", Name: args.Node},
+				}
+				opts := metav1.CreateOptions{}
+				err = m.clientset.CoreV1().Pods(args.PodNamespace).Bind(context.TODO(), binding, opts)
+			}
+		}
+	}
+
+	// return response
+	if err != nil {
+		klog.Error("binding failed:", err.Error())
+		result.Error = err.Error()
+
+		if resourcesAdjusted {
+			// Restore resources to cache. Removing resources should not fail if adding was ok.
+			_ = m.cache.adjustPodResourcesL(pod, remove, annotation, args.Node)
+		}
+	}
+
+	return &result
```

Comment:
does the scheduler recieve a 'bind successful' message from this?

Reply:
if there was an error, then the result will carry that, and bind will fail. see line 432 "result.Error = err.Error()"

When the result doesn't carry an error message, scheduler will consider the return successful.